libs/sql-parser/src/main/java/io/crate/sql/Identifiers.java

```diff
@@ -132,23 +132,41 @@ private static boolean charIsOutsideSafeRange(int i, char c) {
     }
 
     private static boolean quotesRequired(String identifier) {
-        return isKeyWord(identifier) ||
-               // schema names like s, S, "s" are resolved to s (lower case), while "S" (quoted, mixed case) is resolved to S (upper case).
-               // Therefore, mixed case schema names should be quoted whenever necessary. (i.e. printed as a string and to be re-parsed)
-               !IDENTIFIER.matcher(identifier).matches() || isMixedCase(identifier);
+        return isKeyWord(identifier) || !quotesNotRequired(identifier);
     }
 
     public static boolean isKeyWord(String identifier) {
         return RESERVED_KEYWORDS.contains(identifier.toUpperCase(Locale.ENGLISH));
     }
 
-    private static boolean isMixedCase(String identifier) {
-        for (char c : identifier.toCharArray()) {
-            if (c >= 'A' && c <= 'Z') {
-                return true;
+    /**
+     * Quotes are not required only if:
+     *   1) the first char is a lower case or '_' and
+     *   2) the remaining chars are lower cases, digits, or '_'.
+     */
+    private static boolean quotesNotRequired(String identifier) {
+        if (identifier.length() == 0) {
```

Comment:
Can it ever be empty string? maybe an assertion suits better?

Reply:
Thanks for catching, updated it.